x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/assignment/TrainedModelAssignment.java

```diff
@@ -153,6 +156,43 @@ public String[] getStartedNodes() {
             .toArray(String[]::new);
     }
 
+    public Optional<String> selectRandomStartedNodeWeighedOnAllocations() {
+        List<String> nodeIds = new ArrayList<>(nodeRoutingTable.size());
+        List<Integer> cumulativeAllocations = new ArrayList<>(nodeRoutingTable.size());
+        int allocationSum = 0;
+        for (Map.Entry<String, RoutingInfo> routingEntry : nodeRoutingTable.entrySet()) {
+            if (RoutingState.STARTED.equals(routingEntry.getValue().getState())) {
+                nodeIds.add(routingEntry.getKey());
+                allocationSum += routingEntry.getValue().getCurrentAllocations();
+                cumulativeAllocations.add(allocationSum);
+            }
+        }
+
+        if (allocationSum == 0) {
+            return Optional.empty();
+        }
+
+        int randomInt = Randomness.get().ints(1, 1, allocationSum + 1).iterator().nextInt();
+        int nodeIndex = Collections.binarySearch(cumulativeAllocations, randomInt);
+        if (nodeIndex < 0) {
+            nodeIndex = -nodeIndex - 1;
+        }
+        return Optional.of(nodeIds.get(nodeIndex));
+    }
+
+    public Optional<String> selectRandomStartedNodeVersionedBefore(
+        Function<String, Optional<Version>> nodeIdToVersionFunction,
+        Version version
+    ) {
+        List<String> nodeIds = nodeRoutingTable.entrySet()
+            .stream()
+            .filter(e -> e.getValue().getState() == RoutingState.STARTED)
+            .filter(e -> nodeIdToVersionFunction.apply(e.getKey()).orElse(version).before(version))
```

Comment:
This is a non-obvious 'optimization' because we have already checked the "new" nodes and it gave us nothing. But this is assuming a specific behavior from the caller of this method. 

I think either:
 - we don't do this version check
 - We merge this logic into the other function
 - We weigh "old nodes" that are started as having 1 allocation.